### ISSUE TITLE
New open source tables: constant segments

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1147]
**Reported by hansolsson on 29 May 2013 14:19 UTC**
The new constant segments-variant (in Modelica.Blocks.Tables.CombiTable1D) just silently differentiates and this gives incorrect results.

This is a blocking issue if the table is differentiated for index-reduction; since users will get incorrect result without any error indication.

Ideally the failure to differentiate should be detected during index-reduction.

I can see the following solutions (in current Modelica):
* Remove constant segment from this class; possibly have another model for that.
* Use different interpolation routines in the Modelica code (only some of which have derivative-annotation). This will make smoothNess a form of structural parameter.

Using a run-time assertion in the derivative-routine is a possibility - but not a good one, since if the derivative is used for a non-linear system of equations it will generate a Jacobian causing a run-time error; instead of attemping a numeric differentiation.

The model was:
```
model TabTest
  Modelica.Blocks.Sources.Clock clock
    annotation (Placement(transformation(extent={{-80,40},{-60,60}})));
  Modelica.Blocks.Tables.CombiTable1D combiTable1D(table=[0,1; 1,2; 2,4],
      smoothness=Modelica.Blocks.Types.Smoothness.ConstantSegments)
    annotation (Placement(transformation(extent={{-26,40},{-6,60}})));
  Modelica.Blocks.Continuous.Der der1
    annotation (Placement(transformation(extent={{40,40},{60,60}})));
equation 
  connect(clock.y, combiTable1D.u[1]) annotation (Line(
      points={{-59,50},{-28,50}},
      color={0,0,127},
      smooth=Smooth.None));
  connect(combiTable1D.y[1], der1.u) annotation (Line(
      points={{-5,50},{38,50}},
      color={0,0,127},
      smooth=Smooth.None));
  annotation (Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,
            -100},{100,100}}), graphics));
end TabTest;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1147

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 14:34 UTC**
Replying to [ticket:1147 hansolsson]:
> The new constant segments-variant (in Modelica.Blocks.Tables.CombiTable1D) just silently differentiates and this gives incorrect results.

I do not understand.

FYI: We have the same solution in built-in tables of SimulationX and never observed any problems with the derivative set to zero.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 29 May 2013 15:33 UTC**
Replying to [comment:1 beutlich]:
> Replying to [ticket:1147 hansolsson]:
> > The new constant segments-variant (in Modelica.Blocks.Tables.CombiTable1D) just silently differentiates and this gives incorrect results.
> I do not understand.
> 
> FYI: We have the same solution in built-in tables of SimulationX and never observed any problems with the derivative set to zero.

If you take the constant-segments and connect to Der-block and Integrator-block the result should normally be same as the input to the Der-block (except for an offset) - and here the changes in the table are lost.

I'm sure you can also build physical models depending on the derivative being correct. Obviously whether it causes a problem or not depends on how the tables are used.
```
model TestDer4

  Modelica.Blocks.Continuous.Der der1
    annotation (Placement(transformation(extent={{48,40},{68,60}})));
  Modelica.Blocks.Tables.CombiTable1D combiTable1D(table=[0,1; 1,2; 2,4],
      smoothness=Modelica.Blocks.Types.Smoothness.ConstantSegments)
    annotation (Placement(transformation(extent={{-26,40},{-6,60}})));
  Modelica.Blocks.Sources.Clock clock
    annotation (Placement(transformation(extent={{-80,40},{-60,60}})));
  Modelica.Blocks.Continuous.Integrator integrator
    annotation (Placement(transformation(extent={{98,38},{118,58}})));
equation 
  connect(combiTable1D.y[1], der1.u) annotation (Line(
      points={{-5,50},{46,50}},
      color={0,0,127},
      smooth=Smooth.None));
  connect(clock.y, combiTable1D.u[1]) annotation (Line(
      points={{-59,50},{-28,50}},
      color={0,0,127},
      smooth=Smooth.None));
  connect(der1.y, integrator.u) annotation (Line(
      points={{69,50},{82,50},{82,48},{96,48}},
      color={0,0,127},
      smooth=Smooth.None));
  annotation (
    Icon(coordinateSystem(preserveAspectRatio=false, extent={{-100,-100},{100,100}}),
        graphics),
    Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,-100},{100,
            100}}), graphics));
end TestDer4;
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 15:41 UTC**
Replying to [comment:2 hansolsson]:
> If you take the constant-segments and connect to Der-block and Integrator-block the result should normally be same as the input to the Der-block (except for an offset) - and here the changes in the table are lost.

Same holds for step block (Modelica.Blocks.Sources.Step). Step in output signal is lost after der and integration.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 15:51 UTC**
I see, you get an error in Dymola if you differentiate and integrate a step signal.
```
Model error - differentiated if-then-else was not continuous:
 (if time < step.startTime then 0 else step.height)
 Value jumped from 0 to 1.
```
You do not get the error in SimulationX or OM.

So what about the same behaviour for constant segments?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 30 May 2013 07:58 UTC**
BTW: the same problem with differentiation occurs for CombiTimeTable, both with this flag and using the other variant for generating discontinuities:
```
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(smoothness=Modelica.Blocks.Types.Smoothness.LinearSegments,
      table=[0,1; 1,1; 1,2; 2,4])
```

Previously:
* Time-tables had no derivatives and could be discontinuous.
* 1d-tables lookup-tables had derivatives and were continuous

Obviously it would be nice to separate the continuity from the table-type; but still only allow differentiation of continuous tables with continuous outputs.
(Will see if I can find a good example demonstrating that derivatives should integrate correctly.)

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 30 May 2013 13:04 UTC**
Ok, have now constructed a test-example - where I want to use a (simplified) inverse model of a gear-box to find the torque-input that generates the desired speed.

I'm sure it is possible to construct more normal use where tables are for e.g. force-elements inside components and the differentation leads to similar error, since this example is a bit constructed.

The problem is that with the new tables there is no indication that something is incorrect:

```
package Test2
  model M2

    Modelica.Mechanics.Rotational.Components.Inertia inertia(J=1)
      annotation (Placement(transformation(extent={{-44,-10},{-24,10}})));
    Modelica.Mechanics.Rotational.Components.Inertia inertia1(J=1)
      annotation (Placement(transformation(extent={{24,-10},{44,10}})));
    Modelica.Mechanics.Rotational.Sources.Torque torque
      annotation (Placement(transformation(extent={{-84,-10},{-64,10}})));
    Modelica.Mechanics.Rotational.Sensors.SpeedSensor speedSensor
      annotation (Placement(transformation(extent={{54,-10},{74,10}})));
    Modelica.Blocks.Interfaces.RealOutput w1 
      "Absolute angular velocity of flange"
      annotation (Placement(transformation(extent={{92,-8},{112,12}})));
    Modelica.Blocks.Interfaces.RealInput tau1 
      "Accelerating torque acting at flange (= -flange.tau)"
      annotation (Placement(transformation(extent={{-134,-20},{-94,20}})));
    replaceable Modelica.Mechanics.Rotational.Components.IdealGear idealGear(
        ratio=4) constrainedby 
      Modelica.Mechanics.Rotational.Interfaces.PartialElementaryTwoFlangesAndSupport2
      annotation (Placement(transformation(extent={{-14,-10},{6,10}})));
  equation 
    connect(torque.flange, inertia.flange_a) annotation (Line(
        points={{-64,0},{-44,0}},
        color={0,0,0},
        smooth=Smooth.None));
    connect(inertia1.flange_b, speedSensor.flange) annotation (Line(
        points={{44,0},{54,0}},
        color={0,0,0},
        smooth=Smooth.None));
    connect(speedSensor.w, w1) annotation (Line(
        points={{75,0},{102,0},{102,2}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(torque.tau, tau1) annotation (Line(
        points={{-86,0},{-114,0}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(inertia.flange_b, idealGear.flange_a) annotation (Line(
        points={{-24,0},{-14,0}},
        color={0,0,0},
        smooth=Smooth.None));
    connect(idealGear.flange_b, inertia1.flange_a) annotation (Line(
        points={{6,0},{24,0}},
        color={0,0,0},
        smooth=Smooth.None));
    annotation (Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,
              -100},{100,100}}), graphics));
  end M2;

  model IdealTest

    M2 m2_1 annotation (Placement(transformation(
          extent={{-10,-10},{10,10}},
          rotation=180,
          origin={-22,50})));
    Modelica.Blocks.Math.InverseBlockConstraints inverseBlockConstraints
      annotation (Placement(transformation(extent={{-42,38},{-2,62}})));
    M2 m2_2 annotation (Placement(transformation(extent={{20,40},{40,60}})));
    M2 m2_3(redeclare Modelica.Mechanics.Rotational.Components.LossyGear
        idealGear(ratio=4, lossTable=[0,0.9,0.9,0.1,0.1]))
      annotation (Placement(transformation(extent={{20,0},{40,20}})));
    Modelica.Blocks.Tables.CombiTable1D timeTable(table=[0,0; 1,2; 2,3; 3,5],
        smoothness=Modelica.Blocks.Types.Smoothness.ConstantSegments)
      annotation (Placement(transformation(extent={{-80,40},{-60,60}})));
    Modelica.Blocks.Sources.Clock clock
      annotation (Placement(transformation(extent={{-110,40},{-90,60}})));
  equation 
    connect(m2_1.w1, inverseBlockConstraints.u2) annotation (Line(
        points={{-32.2,49.8},{-36,49.8},{-36,50},{-38,50}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(inverseBlockConstraints.y2, m2_1.tau1) annotation (Line(
        points={{-5,50},{-10.6,50}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(inverseBlockConstraints.y1, m2_2.tau1) annotation (Line(
        points={{-1,50},{18.6,50}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(m2_3.tau1, inverseBlockConstraints.y1) annotation (Line(
        points={{18.6,10},{18.6,30},{-1,30},{-1,50}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(clock.y, timeTable.u[1]) annotation (Line(
        points={{-89,50},{-82,50}},
        color={0,0,127},
        smooth=Smooth.None));
    connect(timeTable.y[1], inverseBlockConstraints.u1) annotation (Line(
        points={{-59,50},{-44,50}},
        color={0,0,127},
        smooth=Smooth.None));
    annotation (Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,
              -100},{100,100}}), graphics));
  end IdealTest;
  annotation (uses(Modelica(version="3.2.1")));
end Test2;
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 30 May 2013 13:36 UTC**
Why would you drive a torque with a stepped signal?

Maybe it is possible to check in the table implementation of the derivative calculation if an interval boundary is passed. As long as abscissa is within one interval there should be no step, i.e no. problem.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 30 May 2013 14:41 UTC**
I'm not sure what was behind the question "Why would you drive a torque with a stepped signal?", but I wanted to make something clear from the industrial user's perspective.

Driving systems with step inputs for forces of any kind is a very common things.  At a minimum, it is used to understand the step-response dynamics of a system.  But there are plenty of other reasons why you would get discontinuous forces/torques (e.g. backlash).

Also, I wanted to point out that Hans' use case of building a "backward driven model" (frequently of a drivetrain) is also a common and useful example.  I assume what Hans is pointing out is that normally you cannot specify arbitrary speed profiles because they might be discontinuous.  So normally, you run them through a filter or somehow provide some other assurance that the signal is continuous.  I would certainly expect a warning from the tool if it could not say for certain that it could avoid such discontinuities.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 31 May 2013 13:31 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 31 May 2013 14:16 UTC**
Resolved by ef261103c3932bd8bc1824d19989ac467cb547d6 (after phone call with Martin Otter).
